server/src/main/java/io/crate/expression/scalar/TimeBucketFunction.java

```diff
@@ -0,0 +1,215 @@
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import org.joda.time.Period;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+public class TimeBucketFunction extends Scalar<Long, Object> {
+
+    public static final String NAME = "time_bucket";
+
+    private final Function<Object, Long> intervalOperator;
+
+    // First argument is a result of intervalOperator
+    private final BiFunction<Long, Object, Long> timeBucketOperator;
+
+    // Interval is same for all rows, computed once in compile()
+    // in order to avoid toStandardDuration() call and object allocation per row.
+    @Nullable
+    private Long compiledInterval;
+
+    private enum IntervalOperators {
+        LONG(
+            (Function<Object, Long>) interval -> (Long) interval
+        ),
+        INTERVAL(
+            (Function<Object, Long>) interval -> Long.valueOf(((Period) interval).toStandardDuration().getMillis())
+        );
+
+        private final Function function;
+
+        IntervalOperators(Function function) {
+            this.function = function;
+        }
+
+        public Function getFunction() {
+            return function;
+        }
+    }
+
+    private enum BucketOperators {
+        LONG(
+            (BiFunction<Long, Object, Long>) (interval, timestamp) -> getBucketedTimestamp(interval, (long) timestamp)
+        ),
+        TIMESTAMP(
+            (BiFunction<Long, Object, Long>) (interval, timestamp) -> getBucketedTimestamp(interval, DataTypes.TIMESTAMP.implicitCast(timestamp))
+        ),
+        TIMESTAMPZ(
+            (BiFunction<Long, Object, Long>) (interval, timestamp) -> getBucketedTimestamp(interval, DataTypes.TIMESTAMPZ.implicitCast(timestamp))
+        );
+
+
+        private static long getBucketedTimestamp(long interval, long timestamp) {
+            if (interval <= 0) {
+                throw new IllegalArgumentException("Interval cannot be negative or equal to zero");
+            }
+
+            /*
+             in Java % operator returns negative result only if dividend is negative.
+             https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.17.3
+
+             We need to shift timestamp by the timeline to the "earlier" direction, to the left.
+             But if timestamp is negative (before 1970), remainder will be also negative (see link above)
+             and we need to add it as subtracting would move it to the opposite direction.
+             See test_interval_timestamp_without_zone_works as example.
+            */
+
+            if (timestamp < 0) {
+                return timestamp + timestamp % interval;
+            }
+            return timestamp - timestamp % interval;
```

Comment:
Would this not break with non-fixed-length time units like years, months, ...

Reply:
Interval type is built on top of Period class that explictily restrict month and years and throws an Exception
```
private void checkYearsAndMonths(String destintionType) {
        if (getMonths() != 0) {
            throw new UnsupportedOperationException("Cannot convert to " + destintionType + " as this period contains months and months vary in length");
        }
        if (getYears() != 0) {
            throw new UnsupportedOperationException("Cannot convert to " + destintionType + " as this period contains years and years vary in length");
        }
    }
```
  So time_bucket will show that error. If users want to ignore leap years or treat month as "approximate 30 days period" they need to use either specific days interval or specific milliseconds. I will add this into docs, good point

Reply:
really ? 

```sql
SELECT '2 months 1 week'::INTERVAL
-- 2 mons 1 weeks 00:00:00
```

Reply:
It' possible to have an interval of multiple months or years (it makes sense for date arithmetic when we want to add this interval to some timestamp) but that check is triggered when we want to convert it to duration. 

In general, `time_bucket('2 months':: INTERVAL, timestamp)` doesn't look valid as buckets are different in timeline. 

Should we allow non-equal buckets and in case of month or year do smith like "shift timestamp to the beginning of **it's** month or year"? It's possible with only months or only years interval but smith like 3 _month, 2 weeks, 5 days_ gonna be mess. Probably that why time_bucket docs from above state . _It allows for arbitrary time intervals instead of the second, minute, hour_

Reply:
@mfussenegger, I discussed with @proddata possibility of making origin optional and have some default to make it easier even if in PG [origin is required](https://www.postgresql.org/docs/14/functions-datetime.html#FUNCTIONS-DATETIME-BIN)

I think it makes sense to make it optional as for 1 second, minute, hour, day it doesn't matter what origin we use, it will always be equal to corresponding `date_trunc` result and using default origin would make usage easier.
However, even though it states in docs 
>In the case of full units (1 minute, 1 hour, etc.), it gives the same result as the analogous date_trunc call

it's true only if origin starts on Monday, see [PG test case](https://github.com/postgres/postgres/commit/49ab61f0bdc93984a8d36b602f6f2a15f09ebcc7#diff-96bc081634d3c73577d1ff3900272b287217fd5b7ac095209bd1a1faa04cc688R553).

CrateDB stores timestamp as [epoch milliseconds ](https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/main/java/io/crate/protocols/postgres/types/BaseTimestampType.java#L78) and I think default origin has to be aligned with CrateDB representation i.e DO = 1970-01-01. Another reason is that `date_bin` is just about shifting date to the point where timestamp is divisible by interval, author of date_bin [says](https://www.postgresql.org/message-id/CACPNZCuxeG1mP-6%3DqGJcMpHa8SmHaTXK-zXC8w19HWnu1N-zqA%40mail.gmail.com) about that: 
> It doesn't treat 7 days differently then 5 or 10 without user input (origin or offset), since there is nothing special about 7 day intervals as such internally

And if users want `date_bin` to return Monday  for 1 week, they can provide their custom origin starting from some Monday.

Nit: `date_bin` allows negative intervals. Currently I restrict it but it's not a big deal to allow - just not sure it makes sense, I think it's rather confusing.

Reply:
> I think it makes sense to make it optional as for 1 second, minute, hour, day it doesn't matter what origin we use, it will always be equal to corresponding date_trunc result and using default origin would make usage easier.


If they use 1 second, minute or hour they could also use `date_trunc` directly, no? What do we really gain by making it optional?


> Question is about choosing default origin. CrateDB stores timestamp as epoch milliseconds and I think default origin has to be aligned with CrateDB representation i.e DO = 1970-01-01

How CrateDB stores the timestamp is an implementation detail and shouldn't dictate how functions behave. The semantics of a function should be unrelated to the internal representation of the values.

Reply:
> If they use 1 second, minute or hour they could also use date_trunc directly, no? What do we really gain by making it optional?

I'd say the origin doesn't matter in 99% of the use cases. Having to always supply that, would be not really user friendly (i.e. I would have to remember which year there is a Monday on 1st of january.

> If they use 1 second, minute or hour they could also use date_trunc directly, no? What do we really gain by making it optional?

It imho has nothing to do with single or multiple seconds, minutes, etc.

Reply:
Also the "origin" for `date_trunc()` is not `0` 

```sql
SELECT date_trunc('week',0) ;
--> -259200000
SELECT date_trunc('week',0::timestamp + '1 week'::INTERVAL);
--> 345600000
```

Reply:
>What do we really gain

My fault for bring up date_trunc and unit. In case of any interval with multiple units, say, 15 days, it's still beneficial - A lots of applications store timestamps as epoch milliseconds and when users bin such timestamps they will probably use  1970, Jan 1 as origin - and having it defaulted (and explicitly documented) might be helpful.

Reply:
> A lots of applications treat timestamps as epoch milliseconds and when users bin such timestamps they will probably use 1970, Jan 1 as origin

Sorry, but I strongly disagree, that a user would expect a Thursday to be the start of a **weekly** bucket

Reply:
>origin" for date_trunc() is not 0

date_trunc implementation is special for weeks, it makes sure that day fall to Monday, using ISOChronology.getInstance(tz).weekOfWeekyear()),
and PG also handles week [specially](https://github.com/postgres/postgres/blob/def5b065ff22a16a80084587613599fe15627213/src/backend/utils/adt/timestamp.c#L3896-L3910) 

what I'm saying is that date_bin has different meaning, with rejection of months and years it's just splitting time to equal intervals and putting points to the left ends of the intervals, 7 days is just another argument like 47 days.

Origin is very important for date_bin and it's not even an argument in date_trunc but required in date_bin in PG.
I understand that default origin like 2001-01-01 will make it nice even for 1 week case. Let me summarise all 3 options (required origin, etc epoch 1970 and 2001-01-01) gather some info and I would go with whatever option is voted:

Reply:
As I see it, worst case with having a default value for origin is that PostgreSQL ends up adding a default value later that is different to ours. Then we're either left with a incompatible version or have to do a breaking change. I don't think it is worth the risk.

The explanations given by @BaurzhanSakhariev and the discussion so far also make it seem to me as if forcing the user to choose an origin (and familiarizing themselves with the implications) is a good thing.

Reply:
1. required origin
-  👍 same with PG impl
-  👎 in most cases it would be anyway either Unix time 0 or 01.01.x such that this day is Monday
2. optional origin with default  = unix time 0 (01.01.1970 midnight)
- 👎 not same with PG impl
- 👎 date_trunc(1 week, not_monday_ts) != date_bin(1 week, not_monday_ts) - one should supply own Monday origin to address that but I think then user would simply use date_trunc.
- 👍 GPS trackers use [NMEA](https://en.wikipedia.org/wiki/NMEA_0183) and [RMC for GPS ](http://www.ae.utexas.edu/courses/ase389p7/projects/svatek/commands/RMC.html) has time in UTC and I believe other kinds of sensors use UTC. All calculation are done around UTC zero, as expected.
3. optional origin with default  = 01.01.2001 Monday
- 👎 not same with PG impl
- 👍 date_trunc(1 week, not_monday_ts) == date_bin(1 week, not_monday_ts) 
- 👎 see about sensors and trackers using UTC but here it's disadvantage
 for intervals > 1 day , date_bin can get unexpected results , say for 2 days: (tested locally with PG 14)
```
SELECT date_bin('2 days'::interval, timestamp '1970-01-01 01:00:00' , timestamp '2001-01-01 00:00:00') 
union 
SELECT date_bin('2 days'::interval, timestamp '1970-01-02 01:00:00' , timestamp '2001-01-01 00:00:00') 
```
output, 1 bin, "1970-01-01 00:00:00"

```
SELECT date_bin('2 days'::interval, timestamp '1970-01-01 01:00:00' , timestamp '2001-01-01 00:00:00') 
union 
SELECT date_bin('2 days'::interval, timestamp '1970-01-02 01:00:00' , timestamp '2001-01-01 00:00:00') 
```
output 2 bins
"1969-12-31 00:00:00"
"1970-01-02 00:00:00"

Reply:
Thinking about a bit more a "shortcut" would be to use

```sql
date_bin(<interval>,<timestamp>,0) -- i.e. '1970-01-01 00:00:00'
```
i.e. **option 1**

which would cover most use cases with `hours`, `minutes` and `seconds` (which I'd think are the typical use cases anyway)

Reply:
Origin is a required argument now. With implicit casts, users can use numbers as @proddata mentioned before - but if timestamp is in non-UTC zone, most likely users want origin to be in the same zone and in such cases we need to use `timezone` function. See `test_interval_any_sign_timestamp_with_zone`